src/main/java/io/cryostat/net/reports/ActiveRecordingReportCache.java

```diff
@@ -177,4 +183,27 @@ protected String getReport(RecordingDescriptor recordingDescriptor, String filte
             }
         }
     }
+
+    private void registerConsumer(RecordingDescriptor key) {
+        MessageConsumer<String> consumer = this.vertx.eventBus().consumer(STOP_NOTIFICATION_CATEGORY);
```

Comment:
You can just subscribe to that `STOP_NOTIFICATION_CATEGORY` when this cache class is initialized. In the message handler callback you check if the cache contains the key from the message, and invalidate the cache entry if so. Otherwise you just ignore the message. The message itself should contain some information beyond just the category, so ex. you should be able to get the target and recording name out of it to check if that's in your cache. With the way this is implemented now I think you can run into some effects where the message handler you have registered can end up invalidating a cache entry for a different recording that was stopped, since the `message` contents are not actually checked - you just register the invalidation handler whenever `getReport` is called and then essentially invalidate the corresponding report cache entry the next time any recording is stopped.

Reply:
So if I am understanding it correctly, the `ActiveRecordingReportCache` should implement `NotificationListener` since the class is trying to listen for the `ActiveRecordingStopped` notification, and implement its methods. The categories would be the Set of just that 1 notification category and in the overridden`onMessage`, it would check the cache and maybe the key and recording name for the entry and invalidate it if its there. 

Few questions.

1. What is wrong with publishing in the Notification class like I was doing? Is it because we only care about 1 set of notification categories right now so publishing every other notification would be a waste of resources? 
2. How do I see where the cache class is initialized? I thought the NotificationPublisher would be able to allow each listener implementation class to subscribe to its categories instead of doing that.
3. When I do 
```
public class NotificationPublisher {
    Vertx vertx;
    Set<NotificationListener> listeners;
    
    NotificationPublisher(Vertx vertx, Set<NotificationListener> listeners) {
        listeners.forEach(listener -> {
            listener.categories().forEach(category -> {
                vertx.eventBus().consumer(category).handler(message -> listener.onMessage(category, message));
            });
        });
    }

    <T> void send(String category, T message) {
        vertx.eventBus().publish(category, message);
    }
}
```
Java gives warnings and doesn't like the lack of parameterizations, but I'm not sure how to solve that other than to just put <?> in there. Each listener I assume would use their own generic `T` as a message but the actual 
```
.handler(message -> listener.onMessage(category, message));
            });
```
message callback object is of type Message<Object> so that doesn't help.